src/librustc_driver/args/mod.rs

```diff
@@ -0,0 +1,145 @@
+#![allow(dead_code)]
+
+use std::env;
+use std::error;
+use std::fmt;
+use std::fs;
+use std::io;
+use std::str;
+
+#[cfg(test)]
+mod tests;
+
+/// States for parsing text
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum State {
+    Normal, // within normal text
+    Cr,     // just saw \r
+    Lf,     // just saw \n
+}
+
+struct FileArgs {
+    path: String,
+    input: Vec<u8>,
+    offset: usize,
+}
+
+impl FileArgs {
+    pub fn new(path: String, input: Vec<u8>) -> Self {
+        FileArgs { path, input, offset: 0 }
+    }
+}
+
+impl Iterator for FileArgs {
+    type Item = Result<String, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.input.len() {
+            // All done
+            return None;
+        }
+
+        use State::*;
+        let mut state = Normal;
+        let start = self.offset;
+        let mut end = start;
+
+        for (idx, b) in self.input[start..].iter().enumerate() {
+            let idx = start + idx + 1;
+
+            self.offset = idx;
+
+            match (b, state) {
```

Comment:
This seems somewhat complicated, I think all we really need to do here is split on `\n` and then trim a trailing `\r` if present on each slice, right?

Reply:
Hm, I could be wrong, but maybe we can just do `.lines()` here? I think that does the equivalent unless I'm missing something

Reply:
It would be nice if `.lines()` does the right thing - I'll check it out.

FWIW this code also deals with `\r`-only terminated lines, which is super-obscure at this point (pre-X MacOS) and clearly not justifiable. Also it does something semi-sensible with `\n\r` line breaks, which AFAIK is pointless (I could only imagine it happening if the file had a mix of `\n`-only and `\r\n` line breaks from being mangled by multiple systems).

Reply:
I don't think `.lines` works since this is a `&[u8]` not `&str`, and yeah I think it's fine to not deal with `\r` with a custom state machine parser, splitting on `\n` and optionally trimming `\r` seems reasonable enough.

Reply:
https://doc.rust-lang.org/nightly/std/io/trait.BufRead.html#method.lines is a thing and is implemented for `&[u8]` afaict so we should be able to do it

Reply:
That seems like it's trying a bit too hard, I'd recommend just using "find the newline byte" in a buffer and don't bring I/O into it.